ocaml/xapi/xapi_session.ml

```diff
@@ -661,6 +667,52 @@ let login_no_password_common ~__context ~uname ~originator ~host ~pool
   ignore (Client.Pool.get_all ~rpc ~session_id) ;
   session_id
 
+let login_no_password_common ~__context ~uname ~originator ~host ~pool
+    ~is_local_superuser ~subject ~auth_user_sid ~auth_user_name
+    ~rbac_permissions ~db_ref ~client_certificate =
+  Context.with_tracing ~originator ~__context __FUNCTION__ @@ fun __context ->
+  let internal_xapi_master_to_xapi_slave_login =
+    (xapi_internal_originator, true, true, None)
+  in
+  let is_valid_session session_id =
+    try
+      (* Call an API function to check the session is still valid *)
+      let rpc = Helpers.make_rpc ~__context in
+      ignore (Client.Pool.get_all ~rpc ~session_id) ;
+      true
+    with Api_errors.Server_error (err, _) ->
+      info "Invalid session: %s" err ;
+      false
+  in
+  let create_session () =
+    let new_session_id =
+      login_no_password_common_create_session ~__context ~uname ~originator
+        ~host ~pool ~is_local_superuser ~subject ~auth_user_sid ~auth_user_name
+        ~rbac_permissions ~db_ref ~client_certificate
+    in
+    new_session_id
+  in
+  if
+    (originator, pool, is_local_superuser, uname)
+    = internal_xapi_master_to_xapi_slave_login
+    && !Xapi_globs.reuse_pool_sessions
+  then (
+    if
+      !reusable_pool_session <> Ref.null
+      && ((not !Xapi_globs.validate_reusable_pool_session)
+         || is_valid_session !reusable_pool_session
+         )
+    then
+      !reusable_pool_session
+    else
+      let new_session_id = create_session () in
+      with_lock reusable_pool_session_lock (fun () ->
+          reusable_pool_session := new_session_id
```

Comment:
If we use a lock for writing we also need to use a lock for reading, otherwise this is not different from just having the assignment directly (at least on OCaml 4).

Usually the lock needs to be placed around the whole read-modify-write cycle.

Reply:
Thanks! I do wonder if the performance will be worse though if the threads have to wait for each other to get a session?